#### <font color="#03997A"> Diplomado en Ciencia de datos UNAM </font>

##### <font color="#03997A"> MI: Luis Enrique Aorta V.  </font>

##### <font color="#03997A"> Modulo 9 Natural Language Procesing </font>

##### <font color="#03997A"> Tarea 3 - Procesamiento de Lenguaje Natural
##### <font color="#03997A"> Equipo 1:
- Sergio Ibarra Ramírez
- Yafte Aaron Flores Morales
- Esteban Bracamontes Orozco

</font>

In [ ]:
# Instalamos la libreria
%pip install nltk

In [ ]:
# Importamos la libreria
import nltk
nltk.download('punkt')
nltk.download('wordnet')
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


True

In [ ]:
# Leemos el archivo txt
with open('ElRamoAzul.txt', 'r', encoding='utf-8') as f:
    libro = f.read()
    print(libro[0:1500])

El ramo azul, un cuento de Octavio Paz

Desperté, cubierto de sudor. Del piso de ladrillos rojos, recién regados, subía un vapor caliente. Una mariposa de alas grisáceas revoloteaba encandilada alrededor del foco amarillento. Salté de la hamaca y descalzo atravesé el cuarto, cuidando no pisar algún alacrán salido de su escondrijo a tomar el fresco. Me acerqué al ventanillo y aspiré el aire del campo. Se oía la respiración de la noche, enorme, femenina. Regresé al centro de la habitación, vacié el agua de la jarra en la palangana de peltre y humedecí la toalla. Me froté el torso y las piernas con el trapo empapado, me sequé un poco y, tras de cerciorarme que ningún bicho estaba escondido entre los pliegues de mi ropa, me vestí y calcé. Bajé saltando la escalera pintada de verde. En la puerta del mesón tropecé con el dueño, sujeto tuerto y reticente. Sentado en una sillita de tule, fumaba con el ojo entrecerrado. Con voz ronca me preguntó:
-¿Dónde va señor?

-A dar una vuelta. Hace mucho

In [ ]:
# Checamos la cantidad de caracteres en libro
len(libro)

4705

## A)    ¿Cuántas palabras hay en el texto?

In [ ]:
# Usaremos expresiones regulares para eliminar signos de puntuación y contar las palabras
import re

libro_limpio = re.sub(r'[^\w\s]','',libro)
print(libro_limpio[0:1500])

El ramo azul un cuento de Octavio Paz

Desperté cubierto de sudor Del piso de ladrillos rojos recién regados subía un vapor caliente Una mariposa de alas grisáceas revoloteaba encandilada alrededor del foco amarillento Salté de la hamaca y descalzo atravesé el cuarto cuidando no pisar algún alacrán salido de su escondrijo a tomar el fresco Me acerqué al ventanillo y aspiré el aire del campo Se oía la respiración de la noche enorme femenina Regresé al centro de la habitación vacié el agua de la jarra en la palangana de peltre y humedecí la toalla Me froté el torso y las piernas con el trapo empapado me sequé un poco y tras de cerciorarme que ningún bicho estaba escondido entre los pliegues de mi ropa me vestí y calcé Bajé saltando la escalera pintada de verde En la puerta del mesón tropecé con el dueño sujeto tuerto y reticente Sentado en una sillita de tule fumaba con el ojo entrecerrado Con voz ronca me preguntó
Dónde va señor

A dar una vuelta Hace mucho calor
Hum todo está ya cerrad

In [ ]:
# Tokenizar un texto, produce una lista de palabras y puntuaciones
tokens = nltk.word_tokenize(libro_limpio, language='spanish')
print( type(tokens), len(tokens) )
print( tokens )

<class 'list'> 826
['El', 'ramo', 'azul', 'un', 'cuento', 'de', 'Octavio', 'Paz', 'Desperté', 'cubierto', 'de', 'sudor', 'Del', 'piso', 'de', 'ladrillos', 'rojos', 'recién', 'regados', 'subía', 'un', 'vapor', 'caliente', 'Una', 'mariposa', 'de', 'alas', 'grisáceas', 'revoloteaba', 'encandilada', 'alrededor', 'del', 'foco', 'amarillento', 'Salté', 'de', 'la', 'hamaca', 'y', 'descalzo', 'atravesé', 'el', 'cuarto', 'cuidando', 'no', 'pisar', 'algún', 'alacrán', 'salido', 'de', 'su', 'escondrijo', 'a', 'tomar', 'el', 'fresco', 'Me', 'acerqué', 'al', 'ventanillo', 'y', 'aspiré', 'el', 'aire', 'del', 'campo', 'Se', 'oía', 'la', 'respiración', 'de', 'la', 'noche', 'enorme', 'femenina', 'Regresé', 'al', 'centro', 'de', 'la', 'habitación', 'vacié', 'el', 'agua', 'de', 'la', 'jarra', 'en', 'la', 'palangana', 'de', 'peltre', 'y', 'humedecí', 'la', 'toalla', 'Me', 'froté', 'el', 'torso', 'y', 'las', 'piernas', 'con', 'el', 'trapo', 'empapado', 'me', 'sequé', 'un', 'poco', 'y', 'tras', 'de', 'cerci

### Se tienen 826 palabras en el texto

## B) ¿Cuántas palabras diferentes existen?

In [ ]:
# Eliminemos los duplicado en tokens
palabras_unicas = list(set(tokens))
print(palabras_unicas)

['Arrodíllese', 'en', 'dejas', 'hora', 'rozar', 'una', 'Paz', 'palangana', 'tenso', 'toalla', 'Mi', 'iluminando', 'yo', 'ese', 'ver', 'pocos', 'aquí', 'conversación', 'calle', 'puntas', 'levantándome', 'arrojé', 'rato', 'describió', 'hinqué', 'jardín', 'azules', 'enorme', 'Los', 'cigarrillo', 'Sopló', 'Respiré', 'piedras', 'vasto', 'salido', 'espalda', 'En', 'Ah', 'hay', 'ojos', 'escondido', 'instante', 'Abrí', 'acercaba', 'hamaca', 'hasta', 'muro', 'brillaba', 'ahora', 'haga', 'Alcé', 'trapo', 'cubierto', 'vapor', 'entrecerrar', 'tengo', 'atravesé', 'recién', 'amarillos', 'hizo', 'acodé', 'bien', 'pies', 'desprendía', 'llama', 'cosa', 'no', 'un', 'torso', 'sirven', 'novia', 'alumbrado', 'Pero', 'Mira', 'tanta', 'son', 'vi', 'Se', 'Caminé', 'quemaba', 'estrellas', 'está', 'luminosa', 'lo', 'junto', 'cristiano', 'salió', 'frágil', 'grillo', 'pisar', 'al', 'entrecerrado', 'decir', 'cara', 'dureza', 'Cerré', 'otro', 'incorporé', 'reticente', 'podía', 'Cuando', 'tuerto', 'convenciste', 'va

In [ ]:
len(palabras_unicas)

466

### Se tienen 466 palabras diferentes

## C) ¿Qué cantidad de sustantivos, adjetivos y verbos posee el texto?

#### C.1. Usando metodo NLTK

In [ ]:
# Aplicamos la función POS a la lista de tokens
tokens_pos = nltk.pos_tag(tokens)

In [ ]:
# Revisemos los tipos de tokens encontrados
for i in range(0,len(tokens_pos)):
  print(tokens_pos[i])

('El', 'NNP')
('ramo', 'NN')
('azul', 'NN')
('un', 'JJ')
('cuento', 'NN')
('de', 'IN')
('Octavio', 'NNP')
('Paz', 'NNP')
('Desperté', 'NNP')
('cubierto', 'NN')
('de', 'FW')
('sudor', 'FW')
('Del', 'NNP')
('piso', 'FW')
('de', 'FW')
('ladrillos', 'FW')
('rojos', 'FW')
('recién', 'FW')
('regados', 'FW')
('subía', 'FW')
('un', 'JJ')
('vapor', 'NN')
('caliente', 'NN')
('Una', 'NNP')
('mariposa', 'NN')
('de', 'FW')
('alas', 'FW')
('grisáceas', 'NNS')
('revoloteaba', 'VBP')
('encandilada', 'JJ')
('alrededor', 'NN')
('del', 'NN')
('foco', 'NN')
('amarillento', 'IN')
('Salté', 'NNP')
('de', 'FW')
('la', 'FW')
('hamaca', 'FW')
('y', 'FW')
('descalzo', 'FW')
('atravesé', 'FW')
('el', 'FW')
('cuarto', 'FW')
('cuidando', 'NN')
('no', 'DT')
('pisar', 'NN')
('algún', 'RB')
('alacrán', 'JJ')
('salido', 'NN')
('de', 'IN')
('su', 'FW')
('escondrijo', 'FW')
('a', 'DT')
('tomar', 'NN')
('el', 'NN')
('fresco', 'VB')
('Me', 'NNP')
('acerqué', 'JJ')
('al', 'NN')
('ventanillo', 'NN')
('y', 'NN')
('aspiré', '

In [ ]:
# Contando cuantos sustantivos, adjetivos y verbos se tienen
sustantivos = 0
adjetivos = 0
verbos = 0
for i in range(0,len(tokens_pos)):
  if (tokens_pos[i][1]=='NNP'):
    sustantivos += 1
  if (tokens_pos[i][1]=='NN'):
    adjetivos += 1
  if (tokens_pos[i][1]=='VBZ'):
    verbos += 1

print('El número de sustantivos es: ', sustantivos)
print('El número de adjetivos es: ', adjetivos)
print('El número de verbos es: ', verbos)

El número de sustantivos es:  107
El número de adjetivos es:  349
El número de verbos es:  29


#### C.2. Usando metodo SPACY

In [ ]:
# Haciendo el mismo ejercicio con Spacy
import spacy
!pip install -U pip setuptools wheel
!pip install -U spacy
!python -m spacy download es
!python -m spacy download es_core_news_md

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 10.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 805.1/805.1 kB 15.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.8/64.8 kB 7.0 MB/s eta 0:00:00
  Attempting uninstall: wheel
    Found existing installation: wheel 0.41.1
    Uninstalling wheel-0.41.1:
      Successfully uninstalled wheel-0.41.1
  Attempting uninstall: setuptools
    Found existing installation: setuptools 67.7.2
    Uninstalling setuptools-67.7.2:
      Successfully uninstalled setuptools-67.7.2
  Attempting uninstall: pip
    Found existing installation: pip 23.1.2
    Uninstalling pip-23.1.2:
      Successfully uninstalled pip-23.1.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
ipython 7.34.0 requires jedi>=0.16, which is not installed.


2023-08-23 05:19:37.830224: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
⚠ As of spaCy v3.0, shortcuts like 'es' are deprecated. Please use the
full pipeline package name 'es_core_news_sm' instead.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.9/12.9 MB 22.9 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('es_core_news_sm')
2023-08-23 05:19:54.450773: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.3/42.3 MB 11.5 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('es_core_news_md')


In [ ]:
from spacy.lang.es.examples import sentences

nlp = spacy.load('es_core_news_sm')
doc = nlp(libro)
print(doc.text)
tipos_token = []
for token in doc:
    print(token.text, token.pos_, token.dep_)
    tipos_token.append(token.pos_)

El ramo azul, un cuento de Octavio Paz

Desperté, cubierto de sudor. Del piso de ladrillos rojos, recién regados, subía un vapor caliente. Una mariposa de alas grisáceas revoloteaba encandilada alrededor del foco amarillento. Salté de la hamaca y descalzo atravesé el cuarto, cuidando no pisar algún alacrán salido de su escondrijo a tomar el fresco. Me acerqué al ventanillo y aspiré el aire del campo. Se oía la respiración de la noche, enorme, femenina. Regresé al centro de la habitación, vacié el agua de la jarra en la palangana de peltre y humedecí la toalla. Me froté el torso y las piernas con el trapo empapado, me sequé un poco y, tras de cerciorarme que ningún bicho estaba escondido entre los pliegues de mi ropa, me vestí y calcé. Bajé saltando la escalera pintada de verde. En la puerta del mesón tropecé con el dueño, sujeto tuerto y reticente. Sentado en una sillita de tule, fumaba con el ojo entrecerrado. Con voz ronca me preguntó:
-¿Dónde va señor?

-A dar una vuelta. Hace mucho

In [ ]:
print(tipos_token)

['DET', 'NOUN', 'ADJ', 'PUNCT', 'DET', 'NOUN', 'ADP', 'PROPN', 'PROPN', 'SPACE', 'PROPN', 'PUNCT', 'ADJ', 'ADP', 'NOUN', 'PUNCT', 'ADP', 'NOUN', 'ADP', 'NOUN', 'ADJ', 'PUNCT', 'ADV', 'ADJ', 'PUNCT', 'VERB', 'DET', 'NOUN', 'ADJ', 'PUNCT', 'DET', 'NOUN', 'ADP', 'NOUN', 'ADJ', 'VERB', 'ADJ', 'ADV', 'ADP', 'ADJ', 'NOUN', 'PUNCT', 'NOUN', 'ADP', 'DET', 'NOUN', 'CCONJ', 'PROPN', 'PROPN', 'DET', 'ADJ', 'PUNCT', 'VERB', 'ADV', 'VERB', 'DET', 'NOUN', 'ADJ', 'ADP', 'DET', 'NOUN', 'ADP', 'VERB', 'DET', 'NOUN', 'PUNCT', 'PRON', 'VERB', 'ADP', 'NOUN', 'CCONJ', 'VERB', 'DET', 'NOUN', 'ADP', 'NOUN', 'PUNCT', 'PRON', 'VERB', 'DET', 'NOUN', 'ADP', 'DET', 'NOUN', 'PUNCT', 'ADJ', 'PUNCT', 'ADJ', 'PUNCT', 'PROPN', 'ADP', 'NOUN', 'ADP', 'DET', 'NOUN', 'PUNCT', 'VERB', 'DET', 'NOUN', 'ADP', 'DET', 'NOUN', 'ADP', 'DET', 'NOUN', 'ADP', 'PROPN', 'CCONJ', 'VERB', 'DET', 'NOUN', 'PUNCT', 'PRON', 'VERB', 'DET', 'NOUN', 'CCONJ', 'DET', 'NOUN', 'ADP', 'DET', 'NOUN', 'ADJ', 'PUNCT', 'PRON', 'VERB', 'DET', 'ADV', 'CC

In [ ]:
print('El número de sustantivos es: ', tipos_token.count('NOUN'))
print('El número de adjetivos es: ', tipos_token.count('ADJ'))
print('El número de verbos es: ', tipos_token.count('VERB'))

El número de sustantivos es:  180
El número de adjetivos es:  73
El número de verbos es:  135


#### C.3. Usando metodo FREELING

In [ ]:
# Haciendo el mismo ejercicio con Freeling
import requests
import pandas as pd

# Archivo a ser enviado
files = {'file': open('ElRamoAzul.txt', 'rb')}

# Parámetros
params = {'outf': 'tagged', 'format': 'json'}
# params = {'outf': 'dep', 'format': 'json'}

# Enviar petición
url = "http://www.corpus.unam.mx/servicio-freeling/analyze.php"
r = requests.post(url, files=files, params=params)
lista_tags_freeling = []   ## Aquí guardaremos los tipos de token (tags)
try:
    # Convertir de formato json
    obj = r.json()

    # Ejemplo, obtener todos los lemas
    for sentence in obj:
        for word in sentence:
            # print (word)
            print (word['lemma'],word['token'],word['tag'])
            lista_tags_freeling.append(word['tag'])
except:
    import traceback
    print( traceback.format_exc() )

el El DA0MS0
ramo ramo NCMS000
azul azul AQ0CS00
, , Fc
uno un DI0MS0
cuento cuento NCMS000
de de SP
octavio Octavio NCMS000
paz Paz NCFS000
despertar Desperté VMIS1S0
, , Fc
cubrir cubierto VMP00SM
de de SP
sudor sudor NCMS000
. . Fp
de De SP
el el DA0MS0
piso piso NCMS000
de de SP
ladrillo ladrillos NCMP000
rojo rojos AQ0MP00
, , Fc
recién recién RG
regar regados VMP00PM
, , Fc
subir subía VMII3S0
uno un DI0MS0
vapor vapor NCMS000
caliente caliente AQ0CS00
. . Fp
uno Una DI0FS0
mariposa mariposa NCFS000
de de SP
ala alas NCFP000
grisáceo grisáceas AQ0FP00
revolotear revoloteaba VMII3S0
encandilar encandilada VMP00SF
alrededor alrededor RG
de de SP
el el DA0MS0
foco foco NCMS000
amarillento amarillento AQ0MS00
. . Fp
saltar Salté VMIS1S0
de de SP
el la DA0FS0
hamaca hamaca NCFS000
y y CC
descalzo descalzo NCMS000
atravesar atravesé VMIS1S0
el el DA0MS0
4 cuarto AO0MS00
, , Fc
cuidar cuidando VMG0000
no no RN
pisar pisar VMN0000
alguno algún DI0MS0
alacrán alacrán NCMS000
salir salido 

In [ ]:
# Lista de tags obtenidos
lista_tags_freeling

['DA0MS0',
 'NCMS000',
 'AQ0CS00',
 'Fc',
 'DI0MS0',
 'NCMS000',
 'SP',
 'NCMS000',
 'NCFS000',
 'VMIS1S0',
 'Fc',
 'VMP00SM',
 'SP',
 'NCMS000',
 'Fp',
 'SP',
 'DA0MS0',
 'NCMS000',
 'SP',
 'NCMP000',
 'AQ0MP00',
 'Fc',
 'RG',
 'VMP00PM',
 'Fc',
 'VMII3S0',
 'DI0MS0',
 'NCMS000',
 'AQ0CS00',
 'Fp',
 'DI0FS0',
 'NCFS000',
 'SP',
 'NCFP000',
 'AQ0FP00',
 'VMII3S0',
 'VMP00SF',
 'RG',
 'SP',
 'DA0MS0',
 'NCMS000',
 'AQ0MS00',
 'Fp',
 'VMIS1S0',
 'SP',
 'DA0FS0',
 'NCFS000',
 'CC',
 'NCMS000',
 'VMIS1S0',
 'DA0MS0',
 'AO0MS00',
 'Fc',
 'VMG0000',
 'RN',
 'VMN0000',
 'DI0MS0',
 'NCMS000',
 'VMP00SM',
 'SP',
 'DP3CSN',
 'NCMS000',
 'SP',
 'VMN0000',
 'DA0MS0',
 'NCMS000',
 'Fp',
 'PP1CS00',
 'VMIS1S0',
 'SP',
 'DA0MS0',
 'NCMS000',
 'CC',
 'VMIS1S0',
 'DA0MS0',
 'NCMS000',
 'SP',
 'DA0MS0',
 'NCMS000',
 'Fp',
 'P00CN00',
 'VMII3S0',
 'DA0FS0',
 'NCFS000',
 'SP',
 'DA0FS0',
 'NCFS000',
 'Fc',
 'AQ0CS00',
 'Fc',
 'AQ0FS00',
 'Fp',
 'VMIS1S0',
 'SP',
 'DA0MS0',
 'NCMS000',
 'SP',
 'DA0FS0',
 '

In [ ]:
# Contador de elementos que empiezan con "NC"
contador_NC = 0
for elemento in lista_tags_freeling:
    if elemento.startswith("NC"):
        contador_NC += 1

In [ ]:
# Contador de elementos que empiezan con "AQ"
contador_ADJ = 0
for elemento in lista_tags_freeling:
    if elemento.startswith("AQ"):
        contador_ADJ += 1

In [ ]:
# Contador de elementos que empiezan con "VM"
contador_VRB = 0
for elemento in lista_tags_freeling:
    if elemento.startswith("VM"):
        contador_VRB += 1

In [ ]:
# Contamos la cantidad de sustantivos, adjetivos y verbos
print('El número de sustantivos es: ', contador_NC)
print('El número de adjetivos es: ', contador_ADJ)
print('El número de verbos es: ', contador_VRB)

El número de sustantivos es:  182
El número de adjetivos es:  49
El número de verbos es:  171


#### Tabla comparativa

In [ ]:
# Construimos la tabla para comparar los métodos
import pandas as pd

tabla = pd.DataFrame()

tabla['NLTK'] = None
tabla['Spacy'] = None
tabla['Freeling'] = None

tabla.loc['Sustantivos'] = [sustantivos,tipos_token.count('NOUN'),contador_NC]
tabla.loc['Adjetivos'] = [adjetivos,tipos_token.count('ADJ'),contador_ADJ]
tabla.loc['Verbos'] = [verbos,tipos_token.count('VERB'),contador_VRB]

tabla

NLTK  Spacy  Freeling
Sustantivos   107    180       182
Adjetivos     349     73        49
Verbos         29    135       171

### De la tabla comparativa observamos que NLTK es el qe menos verbos encuentra, observando los datos notamos que en efecto algunos verbos los clasifica como sustantivos. Spacy y Freeling clasificaciones en el mismo orden, más sustantivos, seguidos de verbos y mucho menos adjetivos. La ventaja de Freeling es que considera diferentes opciones para cada categoría.